<h1 align="center">An Introduction to Machine Learning - 25737</h1>
<h4 align="center">Dr. Sajjad Amini</h4>
<h4 align="center">Sharif University of Technology, Spring 2023</h4>

**Student Name**: Arman Lotfalikhani

**Student ID**: 99109166

# Classification

In this question, we will examine classification algorithms. We do not need to implement the algorithms; instead, we will use the functions available in the `scikit-learn` library to solve the question.

## Importing Libraries

First, we import all required libraries.

**Attention**: You should only use these libraries. Other libraries are not acceptable.

In [2]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB

## Loading Data

In this question, we will use the **Breast Cancer** dataset from `scikit-learn`. You can run the following cell to load the dataset:

In [3]:
data = load_breast_cancer() # loading cancer data

X = data.data # features
Y = data.target # labels 

# printing size of the features and targets
print(X.shape)
print(Y.shape)

(569, 30)
(569,)


Now you should use `train_test_split` function to split dataset into three parts:

- 70% for the training set
- 20% for the validation set
- 10% for the test set

Do this in the cell below.

In [7]:
X_train, Y_train, X_val, Y_val, X_test, Y_test = None, None, None, None, None, None

### START CODE HERE ###
X_train, X_test, Y_train, Y_test=train_test_split(X,Y,test_size=0.1)
X_train, X_val, Y_train, Y_val=train_test_split(X_train,Y_train,test_size=2/9)
print(X_train.shape)
print(X_val.shape)
print(X_test.shape)
### END CODE HERE ###

(398, 30)
(114, 30)
(57, 30)


## Logistic Regression

In this part of the question, you will use the `LogisticRegression` classifier to classify the data. You should change the value of `C`, which is the inverse of the regularization parameter, and find the accuracy on the validation set. Using the validation accuracy, you can find the best value for `C`. Note that you may want to change the solver for faster convergence. The `liblinear` solver is recommended for this problem. You can visit [this link](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) for more information about the `LogisticRegression` classifier.

**Question**: What is the best value of `C`?

**Answer**: Around 0.1 to 0.16. But The change in accuracy is very low for c up to 1, so these values can be used as well.

In [41]:
### START CODE HERE ###
Cvals=[0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1, 0.15, 0.9, 1, 3, 5, 10, 100, 1000]
for c in Cvals:
    model=LogisticRegression(solver='liblinear',C=c)
    model.fit(X_train,Y_train)
    print('c=',c,'score=',model.score(X_val,Y_val))
### END CODE HERE ###

c= 1e-06 score= 0.9385964912280702
c= 1e-05 score= 0.9210526315789473
c= 0.0001 score= 0.9210526315789473
c= 0.001 score= 0.9385964912280702
c= 0.01 score= 0.956140350877193
c= 0.1 score= 0.9649122807017544
c= 0.15 score= 0.9649122807017544
c= 0.9 score= 0.956140350877193
c= 1 score= 0.956140350877193
c= 3 score= 0.956140350877193
c= 5 score= 0.956140350877193
c= 10 score= 0.9649122807017544
c= 100 score= 0.956140350877193
c= 1000 score= 0.956140350877193


Now use your best classifier to calculate the accuracy on the test set.

**Qusetion**: What is the accuracy of this classifier on test set?

**Answer**: 92.98245614035088 %

In [24]:
### START CODE HERE ###
model=LogisticRegression(solver='liblinear',C=0.1)
model.fit(X_train,Y_train)
print(model.score(X_test,Y_test))
### END CODE HERE ###

0.9298245614035088


For this additional part, you can use only the first two features of the dataset to classify the data and then draw the data points and the boundary on a figure. You can use the [tutorial](https://aleksandarhaber.com/solve-classification-problems-in-python-scikit-learn-and-visualize-the-classification-results-machine-learning-tutorial/?utm_source=rss&utm_medium=rss&utm_campaign=solve-classification-problems-in-python-scikit-learn-and-visualize-the-classification-results-machine-learning-tutorial) provided to learn how to do this. You may need to use additional libraries such as `matplotlib` specifically for this part.

In [37]:
### START CODE HERE ###
### END CODE HERE ###

## Naive Bayes

In this part, you will use the `GaussianNB` classifier to classify the data. You should not change the default parameters of this classifier. First, train the classifier on the training set and then find the accuracy of it on the test set.

**Question**: What is the accuracy of this method on test set?

**Answer**: 89.47368421052632 %

In [35]:
### START CODE HERE ###
g_model=GaussianNB()
g_model.fit(X_train,Y_train)
g_model.score(X_test,Y_test)
### END CODE HERE ###

0.8947368421052632

## LDA (Linear Discriminant Analysis)

In this part, you will use the `LinearDiscriminantAnalysis` classifier to classify the data. You should not change the default parameters of this classifier. First, train the classifier on the training set and then find the accuracy of it on the test set.

**Question**: What is the accuracy of this method on test set?

**Answer**: 94.73684210526315 %

In [36]:
### START CODE HERE ###
lda_model=LinearDiscriminantAnalysis()
lda_model.fit(X_train,Y_train)
lda_model.score(X_test,Y_test)
### END CODE HERE ###

0.9473684210526315

## Conclution

**Question**: What is the best method for classifying this dataset? What is the best accuracy on the test set?

**Answer**: LDA is the best classification method, followed by linear regression (the defualt sklearn model has a $\lambda ||w||_2^2$ term) and Gaussian naive Bayes is the worst. The best accuracy is 94.74 %